Kafka/src/Akka.Streams.Kafka/Stages/CommittableConsumerStage.cs

```diff
@@ -0,0 +1,223 @@
+            _consumer.OnPartitionsAssigned -= HandleOnPartitionsAssigned;
+            _consumer.OnPartitionsRevoked -= HandleOnPartitionsRevoked;
+
+            Log.Debug($"Consumer stopped: {_consumer.Name}");
+            _consumer.Dispose();
+
+            base.PostStop();
+        }
+
+        //
+        // Consumer's events
+        //
+
+        private void HandleOnMessage(object sender, ConsumerRecord<K, V> message) => _messagesReceived.Invoke(message);
+
+        private void HandleConsumeError(object sender, ConsumerRecord message)
+        {
+            Log.Error(message.Error.Reason);
+            var exception = new SerializationException(message.Error.Reason);
+            switch (_decider(exception))
+            {
+                case Directive.Stop:
+                    // Throw
+                    _completion.TrySetException(exception);
+                    FailStage(exception);
+                    break;
+                case Directive.Resume:
+                    // keep going
+                    break;
+                case Directive.Restart:
+                    // keep going
+                    break;
+            }
+        }
+
+        private void HandleOnError(object sender, Error error)
+        {
+            Log.Error(error.Reason);
+
+            if (!KafkaExtensions.IsBrokerErrorRetriable(error) && !KafkaExtensions.IsLocalErrorRetriable(error))
+            {
+                var exception = new KafkaException(error);
+                FailStage(exception);
+            }
+        }
+
+        private void HandleOnPartitionsAssigned(object sender, List<TopicPartition> list)
+        {
+            _partitionsAssigned.Invoke(list);
+        }
+
+        private void HandleOnPartitionsRevoked(object sender, List<TopicPartition> list)
+        {
+            _partitionsRevoked.Invoke(list);
+        }
+
+        //
+        // Async callbacks
+        //
+
+        private void MessagesReceived(ConsumerRecord<K, V> message)
+        {
+            var consumer = _consumer;
+            var commitableOffset = new CommitableOffset(
+                () => consumer.Commit(),
+                new PartitionOffset("groupId", message.Topic, message.Partition, message.Offset));
+
+            _buffer.Enqueue(new CommittableMessage<K, V>(message, commitableOffset));
+            if (IsAvailable(_out))
+            {
+                Push(_out, _buffer.Dequeue());
+            }
+        }
+
+        private void PartitionsAssigned(IEnumerable<TopicPartition> partitions)
+        {
+            Log.Debug($"Partitions were assigned: {_consumer.Name}");
+            _consumer.Assign(partitions);
+            assignedPartitions = partitions;
+        }
+
+        private void PartitionsRevoked(IEnumerable<TopicPartition> partitions)
+        {
+            Log.Debug($"Partitions were revoked: {_consumer.Name}");
+            _consumer.Unassign();
+            assignedPartitions = null;
```

Comment:
It's better to convert `partitions` to a `List` or array in order to prevent possible multiple `IEnumarable` evaluations. 
And make `assignedPartitions` a `List` or array or immutable array.